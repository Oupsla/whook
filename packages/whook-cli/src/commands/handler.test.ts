import initHandlerCommand from './handler';
import YError from 'yerror';

describe('handlerCommand', () => {
  const log = jest.fn();
  const promptArgs = jest.fn();
  const $injector = jest.fn();

  beforeEach(() => {
    promptArgs.mockReset();
    log.mockReset();
    $injector.mockReset();
  });

  describe('should work', () => {
    it('with all parameters', async () => {
      $injector.mockResolvedValueOnce({
        putEcho: async ({ body }) => ({
          status: 200,
          body,
        }),
      });
      promptArgs.mockResolvedValueOnce({
        _: ['handler'],
        name: 'putEcho',
        parameters: '{"body": {"echo": "YOLO!"} }',
      });

      const handlerCommand = await initHandlerCommand({
        promptArgs,
        log,
        $injector,
      });

      await handlerCommand();

      expect({
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
      }).toMatchSnapshot();
    });

    it('with handler only', async () => {
      $injector.mockResolvedValueOnce({
        getPing: async ({ body }) => ({
          status: 200,
          body,
        }),
      });
      promptArgs.mockResolvedValueOnce({
        _: ['handler'],
        name: 'getPing',
      });

      const handlerCommand = await initHandlerCommand({
        promptArgs,
        log,
        $injector,
      });

      await handlerCommand();

      expect({
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        injectorCalls: $injector.mock.calls,
      }).toMatchSnapshot();
    });
  });

  describe('should fail', () => {
    it('with non JSON parameters', async () => {
      $injector.mockResolvedValueOnce({
        putEcho: async ({ body }) => ({
          status: 200,
          body,
        }),
      });
      promptArgs.mockResolvedValueOnce({
        _: ['handler'],
        name: 'putEcho',
        parameters: '{"body: {"echo": "YOLO!"} }',
      });

      const handlerCommand = await initHandlerCommand({
        promptArgs,
        log,
        $injector,
      });

      try {
        await handlerCommand();
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          promptArgsCalls: promptArgs.mock.calls,
          logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
          injectorCalls: $injector.mock.calls,
        }).toMatchSnapshot();
      }
    });

    it('with a failing handler', async () => {
      $injector.mockResolvedValueOnce({
        putEcho: async () => {
          throw new YError('E_HANDLER_ERROR');
        },
      });
      promptArgs.mockResolvedValueOnce({
        _: ['handler'],
        name: 'putEcho',
        parameters: '{"body": {"echo": "YOLO!"} }',
      });

      const handlerCommand = await initHandlerCommand({
        promptArgs,
        log,
        $injector,
      });

      try {
        await handlerCommand();
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          promptArgsCalls: promptArgs.mock.calls,
          logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
          injectorCalls: $injector.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });
});
