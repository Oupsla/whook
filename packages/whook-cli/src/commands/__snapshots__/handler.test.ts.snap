// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlerCommand should fail with a failing handler 1`] = `
Object {
  "errorCode": "E_UNEXPECTED_SUCCESS",
  "errorParams": Array [],
  "injectorCalls": Array [
    Array [
      Array [
        "putEcho",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "handler",
      "putEcho",
    ],
    Array [
      "debug",
      "parameters",
      Object {
        "body": Object {
          "echo": "YOLO!",
        },
      },
    ],
    Array [
      "error",
      "Got an error while running the handler.",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
}
`;

exports[`handlerCommand should fail with non JSON parameters 1`] = `
Object {
  "errorCode": "E_BAD_PARAMETERS",
  "errorParams": Array [
    "{\\"body: {\\"echo\\": \\"YOLO!\\"} }",
    "Unexpected token e in JSON at position 10",
  ],
  "injectorCalls": Array [],
  "logCalls": Array [],
  "promptArgsCalls": Array [
    Array [],
  ],
}
`;

exports[`handlerCommand should work with all parameters 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "putEcho",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "handler",
      "putEcho",
    ],
    Array [
      "debug",
      "parameters",
      Object {
        "body": Object {
          "echo": "YOLO!",
        },
      },
    ],
    Array [
      "info",
      "{
  \\"status\\": 200,
  \\"body\\": {
    \\"echo\\": \\"YOLO!\\"
  }
}",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
}
`;

exports[`handlerCommand should work with handler only 1`] = `
Object {
  "injectorCalls": Array [
    Array [
      Array [
        "getPing",
      ],
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "handler",
      "getPing",
    ],
    Array [
      "debug",
      "parameters",
      Object {},
    ],
    Array [
      "info",
      "{
  \\"status\\": 200
}",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
}
`;
